Feature: Add pre-commit trivy hook to scan for .env credentials leaks

nshandra opened the pull request:
### :pushpin: References

-   **Issue:** [Investigate Trivy + GitLeaks](https://app.clickup.com/t/86993pdz8)

### :memo: Implementation

The husky hook will trigger when attempting to commit files, if a string matches the scan rule the commit will fail with a report.

nshandra committed:
feat: add husky pre-commit and trivy rules

nshandra committed:
feat: improve hook to warn if trivy is not installed and skip if no file is staged. Make rule match potential leaks with typos (missing quotes or equal). Update README.

nshandra committed:
feat: switch hook to use the committed content instead of the file

nshandra committed:
chore: revert involuntary change to .env

xurxodev reviewed (changes requested):
Thanks @nshandra 

While the intent to prevent credential leaks is excellent, I have concerns about the current implementation approach. The pre-commit hook requires manual Trivy installation on each developer's machine, which significantly reduces its effectiveness and adoption rate.

## 1. **Manual Installation Requirement Defeats the Purpose**

The current implementation requires each developer to manually install Trivy. If Trivy is not installed, the hook simply skips scanning with a warning:

```4:9:.husky/pre-commit
check_trivy_installed() {
  if ! command -v trivy >/dev/null 2>&1; then
    echo "Warning: Trivy is not installed. Pre-commit scanning will be skipped."
    echo "See: https://trivy.dev/latest/getting-started/installation/"
    exit 0
  fi
}
```

**Problem**: This means secrets can still be committed if developers don't have Trivy installed, which defeats the security purpose. The hook becomes optional rather than mandatory, making it ineffective for preventing credential leaks.

**Suggestion**: Consider using a GitHub Action instead, which ensures consistent scanning for all contributors without requiring local setup.

## 2. **Script issues**
I'm not an expert in shell, so I've requested to an IA to analyse it and this is the result

a) Typo in variable name
```35:35:.husky/pre-commit
COMMITED_FILES=$(git diff --name-only --cached)
```
Variable name has a typo: `COMMITED_FILES` should be `COMMITTED_FILES` (two 't's).

b) Files with spaces in names will break
```15:15:.husky/pre-commit
  for file in $files_to_scan; do
```
The loop will break if any staged file has spaces in its name. The variable expansion without quotes will split on whitespace.

c) Relative path to config file
```24:24:.husky/pre-commit
  trivy fs "$temp_dir" --scanners secret --quiet --secret-config trivy-secret.yaml --exit-code 1
```
The relative path `trivy-secret.yaml` may not work if the hook is executed from a different directory. Should use an absolute path based on the script location.

d) Missing error handling for `mktemp`
```14:14:.husky/pre-commit
  local temp_dir=$(mktemp -d)
```
No validation that `mktemp` succeeded. If it fails, the script will continue with an empty or invalid path.

e) Temporary directory cleanup on errors
If the script fails before line 28, the temporary directory won't be cleaned up, leaving orphaned temp directories.

## 3. **Configuration Issues**

#### a) Inconsistency with README
The README states:
```79:79:README.md
Please note that markdown files are not scanned as they may contain examples of how to use the `.env` files.
```
However, there's no exclusion of markdown files in `trivy-secret.yaml` or the hook script. This is misleading and the documentation doesn't match the implementation.

#### b) Regex could be more strict
```6:6:trivy-secret.yaml
    regex: (?i)(?P<key>VITE_DHIS2_AUTH)\s*(?:=|:)?\s*(?P<secret>.+)
```
The `(?P<secret>.+)` pattern will match anything, including empty values or comments. This could lead to false positives or miss edge cases.

## Recommended Solution: GitHub Action

Instead of requiring manual Pre-commit and manual Trivy installation, I strongly recommend using a **GitHub Action** for secret scanning.

### Advantages:
- **No manual installation required** - Works automatically for all contributors
- *Consistent enforcement** - All PRs are scanned, no exceptions
- **Centralized configuration** - Single source of truth
- **Better visibility** - Results appear in GitHub Security tab
- **No developer friction** - Doesn't slow down local development

### Reference:
Use the official Trivy GitHub Action: https://github.com/aquasecurity/trivy-action

nshandra committed:
fix: typo and files with spaces

nshandra committed:
feat: switch scan to pre-push

nshandra committed:
add workflow trivy scan,

xurxodev reviewed (changes requested):
Thanks @nshandra for the improvements. 

The GitHub Action is a great addition. 

However, there are some issues that need to be addressed before approving the PR.

## What's Working Well

-   GitHub Action implementation
-   Improved hook strictness
-   Fixed typo and files with spaces handling
-   Temporary directory cleanup

## Required Changes (Blocking)

### 1. README and Configuration Inconsistency

**Issue:** The README states that markdown files are not scanned, but there's no exclusion in the configuration.

**Location:**

-   `README.md` line 79: "Please note that markdown files are not scanned..."
-   `trivy-secret.yaml`: No `.md` file exclusion
-   `.husky/trivy-leaks-scan.sh`: No `.md` file filtering

---

### 2. Missing `mktemp` Error Validation

**Issue:** No validation that `mktemp` succeeded. If it fails, the script continues with an invalid path.

**Location:** `.husky/trivy-leaks-scan.sh` line 28


---

## Recommended Changes (Non-blocking)

### 3. Relative Path to Config File

**Issue:** The relative path `trivy-secret.yaml` may fail if the script is executed from a different directory.

**Location:** `.husky/trivy-leaks-scan.sh` line 38


---

### 4. Outdated README Information

**Issue:** The README mentions `.husky/pre-commit` but the hook is now `pre-push`.

**Location:** `README.md` lines 75 and 81


---

### 5. Regex Could Be More Strict

**Issue:** The regex `(?P<secret>.+)` may match empty values or comments, potentially causing false positives.

**Location:** `trivy-secret.yaml` line 6

nshandra committed:
fix: update secrets scan README. Add skip and action info.

nshandra committed:
fix: catch errors for mktemp/mkdir/etc.
Fix relative path for trivy-secret.yaml. Change cleanup to a trap.
Switch the base branch selector to use the branch base before falling back to repo default or main/master.
Cleanup pre-push hook.

nshandra committed:
fix: improve regex rule to avoid false positives

nshandra commented:
### 1. README and Configuration Inconsistency
Markdown are excluded by default, that line in the readme is to warn about it. Sorry for the confusion.

### 2. Missing mktemp Error Validation
Added set -e to catch errors for mktemp/mkdir/etc

### 3. Relative Path to Config File
True, fixed. Also switched the compare point selector to use the branch base before falling back to repo default or main/master.

### 4. Outdated README Information
Right, sorry about that. Also added info about the skip options and the action scan.

### 5. Regex Could Be More Strict

Fixed to not trigger at strings like: `VITE_DHIS2_AUTH =   # TODO:`

xurxodev reviewed (approved):
Thanks @nshandra
